### DISCUSSION TITLE
3.2.3 failed to build with Homebrew LLVM 17.  Worked with LLVM 16

### DISCUSSION THREAD

**Discussion by silent-e** [rbenv/ruby-build#2343] [Build failures]
Just thought I'd report an issue or two I had when installing 3.2.3 on my Intel iMac under Ventura via rbenv.  I have both the `llvm` (version 17) and `llvm@16` brews installed and 3.2.3 was giving me the following error when version 17 was linked.

```
touch yjit/target/release/libyjit.a
linking miniruby
ld: warning: ignoring duplicate libraries: '-lpthread'
building rb_mjit_header.h
builtin_binary.inc updated
compiling builtin.c
rb_mjit_header.h updated
building .ext/include/x86_64-darwin22/rb_mjit_min_header-3.2.3.h
linking static-library libruby.3.2-static.a
merging yjit/target/release/libyjit.a into libruby.3.2-static.a
/var/folders/n7/v28hg7gj76j3jjq31kq1_0_h0000gp/T/ruby-build.20240123124456.41701.AkjLhj/ruby-3.2.3/libyjit
llvm-ar: Unknown command line argument '-x'.  Try: 'llvm-ar -help'
llvm-ar: Did you mean '-help'?
llvm-ar: Not enough positional command line arguments specified!
Must specify at least 2 positional arguments: See: llvm-ar -help
make: *** [libruby.3.2-static.a] Error 1
make: *** Waiting for unfinished jobs....
transform_mjit_header: making external definition static inline: rb_vm_lvar_exposed, Init_vm_stack_canary, rb_vm_opt_cfunc_p, rb_vm_ic_hit_p, rb_vm_opt_newarray_min, rb_vm_opt_newarray_max, rb_vm_splat_array, rb_vm_concat_array, rb_vm_defined, rb_vm_cc_general, rb_aliased_callable_method_entry, rb_find_defined_class_by_owner, rb_vm_call_ivar_attrset_p, rb_iseq_only_kwparam_p, rb_iseq_only_optparam_p, rb_vm_setinstancevariable, rb_vm_setclassvariable, rb_vm_getclassvariable, rb_vm_get_ev_const, rb_vm_rewrite_cref, rb_vm_get_cref, rb_vm_push_frame_fname, rb_vm_pop_frame_no_int, rb_vm_block_ep_update, rb_vm_cref_new_toplevel, rb_vm_cref_dup_without_refinements, rb_vm_frame_block_handler, rb_vm_cframe_keyword_p, rb_vm_ep_local_ep
transform_mjit_header: making declaration static inline: rb_vm_lvar_exposed, rb_vm_lvar_exposed, Init_vm_stack_canary, rb_vm_rewrite_cref, rb_vm_pop_frame_no_int, rb_vm_frame_block_handler, rb_vm_block_ep_update, rb_vm_ep_local_ep, rb_vm_push_frame_fname
transform_mjit_header: SKIPPED to transform: __sigbits, isspecial, isrune, isphonogram, isnumber, isideogram, ishexnumber, digittoint, toupper, tolower, toascii, isxdigit, isupper, isspace, ispunct, isprint, islower, isgraph, isdigit, iscntrl, isblank, isalpha, isalnum, __wcwidth, __isctype, __istype, isascii, __sincospi, __sincospif, __sincos, __sincosf, __inline_isnormall, __inline_isnormald, __inline_isnormalf, __inline_signbitl, __inline_signbitd, __inline_signbitf, __inline_isnanl, __inline_isnand, __inline_isnanf, __inline_isinfl, __inline_isinfd, __inline_isinff, __inline_isfinitel, __inline_isfinited, __inline_isfinitef, __darwin_fd_clr, __darwin_fd_set, __darwin_fd_isset, __darwin_check_fd_set, __sputc
ln -sf ../../../.ext/include/x86_64-darwin22/rb_mjit_min_header-3.2.3.h include/ruby-3.2.0/x86_64-darwin22/rb_mjit_min_header-3.2.3.h
external command failed with status 2
```

However when I switched to llvm 16 using `brew link --overwrite llvm@16` only then did 3.2.3 install.  One other thing I had to do which might be of use to someone was I had to set `RUBY_CONFIGURE_OPTS="rb_cv_symbol_prefix=_"` because of the following error.

```
touch yjit/target/release/libyjit.a
linking miniruby
ld: warning: ignoring duplicate libraries: '-lpthread'
ld: Undefined symbols:
  _coroutine_transfer, referenced from:
      _fiber_setcontext in cont.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [miniruby] Error 1
```

Not sure if both need to be addressed but wanted to mention it for anyone else having these issues.

(e)

**Comment by hsbt** [comment]
I'm not investigating it deeply yet.

You can build it with `brew unlink llvm` or unset `LDFLAGS` and `CPPFLAGS`. Because ruby build system is not working `libtool` and `llvm` toolchains from homebrew. It's only works with `ar` provided by Xcode.

**Comment by sjrmanning** [comment]
edit: removing this assumption as it turned out to be incorrect, re: Xcode latest version and llvm versions.

**Reply by monfresh** [reply]
I just upgraded to macOS 14.4 and command line tools 15.3, and I uninstalled and reinstalled Ruby 3.2.3 without any issues. I'm not linking the Homebrew `llvm`.

When you say "this affects anyone using the latest Xcode", what is "this"? Are you not able to install Ruby at all, or only if you link the Homebrew `llvm`? Homebrew does not link `llvm` by default, so I'm curious if you're linking for a specific reason.

**Reply by sjrmanning** [reply]
I don't have Homebrew llvm installed, but I'm getting the same error as the above (`llvm-ar: Unknown command line argument '-x'.  Try: 'llvm-ar -help'`). But if you were able to do it on a fresh install, most likely there's something else about my setup that is causing this. Will do some more digging.

**Reply by sjrmanning** [reply]
So confirmed this is just a very specific issue to my system. I was able to get around this by linking llvm@16 from brew and completely re-installing Xcode's command-line tools.
